src/bun.js/api/html_rewriter.zig

```diff
@@ -409,59 +407,75 @@ pub const HTMLRewriter = struct {
+                    },
+                };
+            };
+            // sync error occurs
+            if (sink.tmp_sync_error) |err| {
+                sink.tmp_sync_error = null;
+                return err;
             }
 
             // Hold off on cloning until we're actually done.
             return sink.response.toJS(sink.global);
         }
 
-        pub fn onFinishedLoading(sink: *BufferOutputSink, bytes: JSC.WebCore.Blob.Store.ReadFile.ResultType) void {
-            switch (bytes) {
-                .err => |err| {
-                    if (sink.response.body.value == .Locked and @intFromPtr(sink.response.body.value.Locked.task) == @intFromPtr(sink) and
-                        sink.response.body.value.Locked.promise == null)
-                    {
-                        sink.response.body.value = .{ .Empty = {} };
-                        // is there a pending promise?
-                        // we will need to reject it
-                    } else if (sink.response.body.value == .Locked and @intFromPtr(sink.response.body.value.Locked.task) == @intFromPtr(sink) and
-                        sink.response.body.value.Locked.promise != null)
-                    {
-                        sink.response.body.value.Locked.onReceiveValue = null;
-                        sink.response.body.value.Locked.task = null;
-                    }
-
-                    sink.response.body.value.toErrorInstance(err.toErrorInstance(sink.global), sink.global);
-                    sink.rewriter.end() catch {};
-                    sink.deinit();
-                    return;
-                },
-                .result => |data| {
-                    _ = sink.runOutputSink(data.buf, true, data.is_temporary);
-                },
+        pub fn onFinishedBuffering(ctx: *anyopaque, bytes: []const u8, js_err: ?JSC.JSValue, is_async: bool) void {
+            const sink = bun.cast(*BufferOutputSink, ctx);
+            if (js_err) |err| {
+                if (sink.response.body.value == .Locked and @intFromPtr(sink.response.body.value.Locked.task) == @intFromPtr(sink) and
+                    sink.response.body.value.Locked.promise == null)
+                {
+                    sink.response.body.value = .{ .Empty = {} };
+                    // is there a pending promise?
+                    // we will need to reject it
+                } else if (sink.response.body.value == .Locked and @intFromPtr(sink.response.body.value.Locked.task) == @intFromPtr(sink) and
+                    sink.response.body.value.Locked.promise != null)
+                {
+                    sink.response.body.value.Locked.onReceiveValue = null;
+                    sink.response.body.value.Locked.task = null;
+                }
+                if (is_async) {
+                    sink.response.body.value.toErrorInstance(err, sink.global);
+                } else {
+                    sink.tmp_sync_error = throwLOLHTMLError(sink.global);
```

Comment:
this will segfault 

we cannot hold this value in the heap without the GC visiting it

Reply:
fixed in https://github.com/oven-sh/bun/pull/4418/commits/28eafc84e0d67e910bd9af656426aa3cdfba5c12 and https://github.com/oven-sh/bun/pull/4418/commits/e4ae61ac86219e56a766cd85712bd3d1749f7f4c